scipy/integrate/tests/test_quadrature.py

```diff
@@ -290,3 +292,83 @@ def test_trapz(self):
         assert_equal(trapezoid(y, x=x, dx=0.5, axis=0),
                      trapz(y, x=x, dx=0.5, axis=0))
 
+
+class TestQMCQuad():
+    def test_input_validation(self):
+        message = "`func` must be callable."
+        with pytest.raises(TypeError, match=message):
+            qmc_quad("a duck", [[0, 1], [0, 1]])
+
+        message = "`func` must evaluate the integrand at points..."
+        with pytest.raises(ValueError, match=message):
+            qmc_quad(lambda x: 1, [[0, 1], [0, 1]])
+
+        message = "Exception encountered when attempting vectorized call..."
+        # This is OK; function can be vectorized automatically if args is None
+        with pytest.warns(UserWarning, match=message):
+            qmc_quad(lambda x: (x+np.array([1, 2, 3])).sum(), [[0, 1]])
+        # This is not; raise an error
+        message = "Exception encountered when attempting vectorized call..."
+        with pytest.raises(ValueError, match=message):
+            qmc_quad(lambda x, a: x+np.array([1, 2, 3]), [[0, 1]], args=1)
+
+        message = "`n_points` must be an integer."
+        with pytest.raises(TypeError, match=message):
+            qmc_quad(lambda x, y: 1, [[0, 1], [0, 1]], n_points=1024.5)
+
+        message = "`n_offsets` must be an integer."
+        with pytest.raises(TypeError, match=message):
+            qmc_quad(lambda x, y: 1, [[0, 1], [0, 1]], n_offsets=8.5)
+
+        message = "`qrng` must be an instance of scipy.stats.qmc.QMCEngine."
+        with pytest.raises(TypeError, match=message):
+            qmc_quad(lambda x, y: 1, [[0, 1], [0, 1]], qrng="a duck")
+
+        message = "`qrng` must be initialized with dimensionality equal to "
+        with pytest.raises(ValueError, match=message):
+            qmc_quad(lambda x, y: 1, [[0, 1], [0, 1]], qrng=stats.qmc.Sobol(1))
+
+        message = r"`log` must be boolean \(`True` or `False`\)."
+        with pytest.raises(TypeError, match=message):
+            qmc_quad(lambda x, y: 1, [[0, 1], [0, 1]], log=10)
+
+    @pytest.mark.parametrize("n_points", [2**8, 2**12])
+    @pytest.mark.parametrize("n_offsets", [8, 16])
+    def test_basic(self, n_points, n_offsets):
+
+        ndim = 2
+        mean = np.zeros(ndim)
+        cov = np.eye(ndim)
+
+        def func(x, y, mean, cov):
+            z = np.asarray([x, y]).T
+            return stats.multivariate_normal.pdf(z, mean, cov)
+
+        rng = np.random.default_rng(2879434385674690281)
+        qrng = stats.qmc.Sobol(ndim, seed=rng)
+        lb = np.zeros(ndim)
+        ub = np.ones(ndim)
+        ranges = np.asarray([lb, ub]).T
+        res = qmc_quad(func, ranges, n_points=n_points, n_offsets=n_offsets,
+                       args=(mean, cov), qrng=qrng)
+        ref = stats.multivariate_normal.cdf(ub, mean, cov, lower_limit=lb)
```

Comment:
In the [QMC tutorial](https://github.com/scipy/scipy/blob/main/doc/source/tutorial/stats/plots/qmc_plot_conv_mc_sobol.py), there is another function that we could use to check that we do get a the expected convergence rate of QMC (`1/n` in general and `1/(n**3/2)` for Sobol' on good integrands. As opposed to `1/(n**1/2)` for MC).

Some other functions that I often use to test QMC methods:

```python
_exp1 = 1 - np.exp(1)


def art_0(sample):
    # dim 2, true value 0
    return np.prod(sample - 0.5, axis=1)


def art_1d(sample):
    # dim 1, true value 1
    return sample * np.exp(sample)


def art_1(sample):
    # dim 5, true value 0
    return np.sum(np.exp(sample) + _exp1, axis=1)


def art_2(sample):
    # dim 5, true value 5/3 + 5*(5 - 1)/4
    return np.sum(sample, axis=1) ** 2


def art_3(sample):
    # dim 3, true value 0
    return np.prod(np.exp(sample) + _exp1, axis=1)


def type_a(sample, dim=30):
    # true value 1
    a = np.arange(1, dim + 1)
    f = 1.
    for i in range(dim):
        f *= (abs(4. * sample[:, i] - 2) + a[i]) / (1. + a[i])
    return f


def type_b(sample, dim=30):
    # true value 1
    f = 1.
    for d in range(1, dim + 1):
        f *= (d - sample[:, d - 1]) / (d - 0.5)
    return f


def type_b_f2(sample, dim=30):
    # true value 1
    f = 1.
    inv = 1. / dim
    for d in range(dim):
        f *= (1. + 1. / dim) * sample[:, d] ** inv
    return f


def type_c(sample, dim=10):
    # true value 1
    f = 2 ** dim * np.prod(sample, axis=1)
    return f


def additive(sample):
    # true value 1/2*dim
    return np.sum(sample, axis=1)


def pi(x):
    # true value np.pi
    radius = (x[:, 0]**2 + x[:, 1]**2) ** .5
    in_circle = len(np.where(radius <= 1)[0])
    return 4 * in_circle / x.shape[0]


functions = namedtuple('functions', ['name', 'func', 'dim', 'ref'])

benchmark = [
    functions('Art 2D', art_0, 2, 0),
    functions('Art 1D', art_1d, 1, 1),
    functions('Art 1', art_1, 5, 0),
    functions('Art 2', art_2, 5, 5 / 3 + 5 * (5 - 1) / 4),
    functions('Art 3', art_3, 3, 0),
    functions('Type A', type_a, 30, 1),
    functions('Type B f3', type_b, 30, 1),
    functions('Type B f2', type_b_f2, 30, 1),
    functions('Type C', type_c, 10, 1),
    functions('Additive', additive, 30, 1/2*30)
]
```